File: 1-100/5-longest-palindromic-substring.cpp
string longestPalindrome(string s) {
    int n = s.size();
    if (n == 1) return s;
    int maxLength = 1;
    int maxStart = 0;
    vector<vector<bool>> dp(n, vector<bool>(n, false));

    for (int i = 0; i < n; ++i) {
        dp[i][i] = true;
        if (i < n - 1) {
            if (s[i] == s[i+1]) {
                dp[i][i+1] = true;
                maxLength = 2;
                maxStart = i;
            }
        }
    }

    for (int len = 3; len <= n; ++len) {
        for (int i = 0; i + len - 1 < n; ++i) {
            int j = i + len - 1;
            if (s[j] == s[i] && dp[i+1][j-1] == true) {
                dp[i][j] = true;
                maxStart = i;
                maxLength = len;
            }
        }
    }

    return s.substr(maxStart, maxLength);
}

File: python version/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False for _ in range(n)] for _ in range(n)]
        max_len = 1
        max_start = 0
        for i in range(n):
            if i < n-1 and s[i] == s[i+1]:
                dp[i][i+1] = True
                max_len = 2
                max_start = i
            dp[i][i] = True
        
        for s_len in range(3, n+1):
            for i in range(0, n-s_len+1):
                j = i+s_len-1
                if dp[i+1][j-1] and s[i] == s[j]:
                    dp[i][j] = True 
                    if s_len > max_len:
                        max_len = s_len 
                        max_start = i
        return s[max_start:max_start+max_len]
    

# Another way to dp

class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        dp = [[False for _ in range(n)] for _ in range(n)]
        max_len = 1
        max_start = 0
        
        for i in range(n-1, -1, -1):
            for j in range(i, n):
                if j-i <= 1 and s[j] == s[i]:
                    dp[i][j] = True
                    if j-i+1 > max_len:
                        max_len = j-i+1
                        max_start = i
                elif dp[i+1][j-1] and s[j] == s[i]:
                    dp[i][j] = True
                    if j-i+1 > max_len:
                        max_len = j-i+1
                        max_start = i

        return s[max_start:max_start+max_len]